### PR TITLE
cherrypick(prometheus): performance optimization

### PR DESCRIPTION
**Pull request by StarlightIbuki**
cherry-picking 2 commits that optimize the performance.

### CODE REVIEW

File: kong/plugins/prometheus/exporter.lua [hunk 1]
```diff
@@ -315,20 +325,22 @@ local function metric_data()
...
 
-  metrics.connections:set(ngx.var.connections_active or 0, { "active" })
-  metrics.connections:set(ngx.var.connections_reading or 0, { "reading" })
-  metrics.connections:set(ngx.var.connections_writing or 0, { "writing" })
-  metrics.connections:set(ngx.var.connections_waiting or 0, { "waiting" })
+  if http_subsystem then -- only export those metrics once in http as they are shared
+    metrics.connections:set(ngx.var.connections_active or 0, { "active" })
+    metrics.connections:set(ngx.var.connections_reading or 0, { "reading" })
+    metrics.connections:set(ngx.var.connections_writing or 0, { "writing" })
+    metrics.connections:set(ngx.var.connections_waiting or 0, { "waiting" })
```
Comment (StarlightIbuki) [inline kong/plugins/prometheus/exporter.lua:332]: Not sure if I'm doing this right.

File: kong/plugins/prometheus/api.lua [hunk 2]
```diff
@@ -1,13 +1,20 @@
-local prometheus = require "kong.plugins.prometheus.exporter"
+local exporter = require "kong.plugins.prometheus.exporter"
 
 local printable_metric_data = function()
-  return table.concat(prometheus.metric_data(), "")
+  local buffer = {}
+  -- override write_fn, since stream_api expect response to returned
+  -- instead of ngx.print'ed
+  exporter.metric_data(function(data)
+    table.insert(buffer, table.concat(data, ""))
```
Comment (chronolaw) [inline kong/plugins/prometheus/api.lua:8]: should we localize `table.insert` and `table.concat`?

File: kong/plugins/prometheus/exporter.lua [hunk 3]
```diff
@@ -33,6 +33,7 @@ if pok then
 end
 
 local kong_subsystem = ngx.config.subsystem
+local http_subsystem = kong_subsystem == "http"
```
Comment (chronolaw) [inline kong/plugins/prometheus/exporter.lua:36]: should be `is_http_subsystem`

File: kong/plugins/prometheus/exporter.lua [hunk 4]
```diff
@@ -47,31 +48,35 @@ local function init()
   if kong_subsystem == "http" then
```
Comment (chronolaw) [inline kong/plugins/prometheus/exporter.lua:48]: use `is_http_subsystem`

File: kong/plugins/prometheus/exporter.lua [hunk 5]
```diff
@@ -315,20 +325,22 @@ local function metric_data()
     end
   end
 
-  metrics.connections:set(ngx.var.connections_active or 0, { "active" })
-  metrics.connections:set(ngx.var.connections_reading or 0, { "reading" })
-  metrics.connections:set(ngx.var.connections_writing or 0, { "writing" })
-  metrics.connections:set(ngx.var.connections_waiting or 0, { "waiting" })
+  if http_subsystem then -- only export those metrics once in http as they are shared
+    metrics.connections:set(ngx.var.connections_active or 0, { "active" })
```
Comment (chronolaw) [inline kong/plugins/prometheus/exporter.lua:329]: localize `ngx.var`

File: kong/plugins/prometheus/api.lua (hunk 2 above)
Comment (StarlightIbuki) [reply kong/plugins/prometheus/api.lua:8]: This is a cherry-pick.

File: kong/plugins/prometheus/exporter.lua (hunk 4 above)
Comment (StarlightIbuki) [reply kong/plugins/prometheus/exporter.lua]: done

File: kong/plugins/prometheus/exporter.lua (hunk 3 above)
Comment (StarlightIbuki) [reply kong/plugins/prometheus/exporter.lua:36]: Ditto.

File: kong/plugins/prometheus/exporter.lua (hunk 5 above)
Comment (StarlightIbuki) [reply kong/plugins/prometheus/exporter.lua:329]: Ditto.

**Review by fffonion — APPROVED** [review]

**Pull request merged**